### ISSUE TITLE
How to implement a custom converter ( as part of a sklearn pipeline) for mapping values

### ISSUE THREAD

**Issue by waiyujack** [onnx/sklearn-onnx#557]
I am trying to code up a custom converter that maps values say non-positives to -1 and positives to 1.

Something like below in terms of python code:
```
def mapping(x):
    return np.where(x <= 0, -1, 1)
```

I am thinking of using the OnnxIF function but haven't got much idea of how to use that as there aren't any examples. Can anyone help me? Other methods are also welcome!

I have been trying to replicate the simple scaler example from [here ](http://onnx.ai/sklearn-onnx/auto_examples/plot_convert_syntax.html#pipeline-and-a-custom-object) but haven't got much success so far. I am struggling to write some code on the last line of **class CustomOpTransformer**.

```

# **Imports**

import numpy as np
from sklearn.base import BaseEstimator, TransformerMixin
from sklearn.pipeline import make_pipeline
from onnxruntime import InferenceSession
from skl2onnx import convert_sklearn, to_onnx, wrap_as_onnx_mixin
from skl2onnx.common.data_types import FloatTensorType
from skl2onnx.algebra.onnx_operator_mixin import OnnxOperatorMixin

# **Simple function to check the converted model works fine.**

def predict_with_onnxruntime(onx, X):
    sess = InferenceSession(onx.SerializeToString())
    input_name = sess.get_inputs()[0].name
    res = sess.run(None, {input_name: X.astype(np.float32)})
    return res[0]

# **Custom transformer to perform mapping (NOT YET COMPLETE see last line)**

class CustomOpTransformer(BaseEstimator, TransformerMixin,
                          OnnxOperatorMixin):

    def __init__(self):
        BaseEstimator.__init__(self)
        TransformerMixin.__init__(self)

    def fit(self, X, y=None):
        self.W_ = np.float32(1)
        self.S_ = np.float32(1)
        return self

    def transform(self, X):
        return np.where(X <= 0, -1, 1)

    def onnx_shape_calculator(self):
        def shape_calculator(operator):
            operator.outputs[0].type = operator.inputs[0].type
        return shape_calculator
    
    
    # We need to use some if else statments
    def to_onnx_operator(self, inputs=None, outputs=('Y', )):
        if inputs is None:
            raise RuntimeError("Parameter inputs should contain at least "
                               "one name.")
        i0 = self.get_inputs(inputs, 0)
        return **????**

# **Try to implement the convertor in a pipeline and test the function**

X = np.array([[0, 0], [1, 1]])
tr = make_pipeline(CustomOpTransformer())
tr.fit(X)

# Original matrix
print('Original matrix')
print(X)

# sklearn pipeline
print('Transform using np.where')
print(tr.transform(X))

# Run in onnx
print('Transform using to_onnx_operator')
onx = to_onnx(tr, X.astype(np.float32), target_opset=12)
print(predict_with_onnxruntime(onx, X))
```

**Comment by xadupre** [comment]
Hi,

I would do it that way:

```
# **Imports**

import numpy as np
from sklearn.base import BaseEstimator, TransformerMixin
from sklearn.pipeline import make_pipeline
from onnxruntime import InferenceSession
from skl2onnx import convert_sklearn, to_onnx, wrap_as_onnx_mixin
from skl2onnx.common.data_types import FloatTensorType
from skl2onnx.algebra.onnx_operator_mixin import OnnxOperatorMixin
from skl2onnx.algebra.onnx_ops import OnnxWhere, OnnxLessOrEqual
from skl2onnx.common.data_types import guess_numpy_type

# **Simple function to check the converted model works fine.**

def predict_with_onnxruntime(onx, X):
    sess = InferenceSession(onx.SerializeToString())
    input_name = sess.get_inputs()[0].name
    res = sess.run(None, {input_name: X.astype(np.float32)})
    return res[0]

# **Custom transformer to perform mapping (NOT YET COMPLETE see last line)**

class CustomOpTransformer(BaseEstimator, TransformerMixin,
                          OnnxOperatorMixin):

    def __init__(self):
        BaseEstimator.__init__(self)
        TransformerMixin.__init__(self)

    def fit(self, X, y=None):
        self.W_ = np.float32(1)
        self.S_ = np.float32(1)
        return self

    def transform(self, X):
        return np.where(X <= 0, -1, 1)

    def onnx_shape_calculator(self):
        def shape_calculator(operator):
            operator.outputs[0].type = operator.inputs[0].type
        return shape_calculator
    
    
    # We need to use some if else statments
    def to_onnx_operator(self, inputs=None, outputs=('Y', )):
        if inputs is None:
            raise RuntimeError("Parameter inputs should contain at least "
                               "one name.")
        i0 = self.get_inputs(inputs, 0)
        dtype = guess_numpy_type(i0.type)
        return OnnxWhere(
                OnnxLessOrEqual(i0, np.array([0], dtype), op_version=12),
                np.array([-1], dtype), np.array([1], dtype),
                op_version=12, output_names=outputs[0])
        

# **Try to implement the convertor in a pipeline and test the function**

X = np.array([[0, 0], [1, 1]])
tr = make_pipeline(CustomOpTransformer())
tr.fit(X)

# Original matrix
print('Original matrix')
print(X)

# sklearn pipeline
print('Transform using np.where')
print(tr.transform(X))

# Run in onnx
print('Transform using to_onnx_operator')
onx = to_onnx(tr, X.astype(np.float32), target_opset=12)
print(predict_with_onnxruntime(onx, X))
```

**Comment by waiyujack** [comment]
Thanks xadupre! I think your solution would just work on version 1.7.1 right?

Do you mind if I also extend the question a bit? Instead of mapping from intervals to integers, do you know if Onnx operator would take dictionary formats and do one to one mapping (integer to integer)? Could you provide some example of how it would take a dictionary variable. A simplified numpy example is given below:

```
X = np.array([[0, 0], [1, 1]])
dictionary = {0:-1, 1:1}

# define a function for mapping
def mapping(x):
    return(dictionary[x])

# vectorise my function for matrix mapping
mapping_vec = np.vectorize(mapping)

mapping_vec(X)
```

The reason I asked this is because I am planning to use _different_ mappings on a few _different_ features before predicting/modelling.

**Comment by waiyujack** [comment]
Hi @xadupre 

I have been comparing the speed of the above example between sklearn and onnxruntime.

It seems to me the onnx operators approach is about 6 times slower than running it just using sklearn pipline. I am wondering if you have any thoughts on this and whether feature mapping could be performed quicker within onnxruntime?

Script for comparing the sklearn pipeline and onnxruntime for the above example:

```
from timeit import timeit

# Generate a random matrix 
X = np.random.rand(1000,1000)
tr = make_pipeline(CustomOpTransformer())
tr.fit(X)
onx = to_onnx(tr, X.astype(np.float32), target_opset=12)

# Comparison

repeats = 1000

print("Predicting on data frame with dimensions: " + str(X.shape) + ' ' + str(repeats) + ' times')

print("Sklearn: " + str( timeit("tr.transform(X)", number=repeats, globals=globals())) + ' s')

print("ONNX: " + str( timeit("predict_with_onnxruntime(onx, X)", number=repeats, globals=globals())) + ' s')
```

Results:
```
Predicting on data frame with dimensions: (1000, 1000) 1000 times
Sklearn: 1.6012675259262323 s
ONNX: 10.493801097851247 s
```

**Comment by waiyujack** [comment]
Hi @xadupre 

I just want to see if you have any thoughts on the performance on onnxwhere vs np.where. Do you think it's better to leave custom mappings outside onnx runtime?

**Comment by xadupre** [comment]
It depends on your model. I think it is better to keep as much as as possible inside onnxruntime. There is room for improvments in the implementation of Where in onnxruntime (CPU). I did a benchmark between onnxruntime, numpy, tensorflow, pytorch for operator where: [Compares implementations of Where](http://www.xavierdupre.fr/app/mlprodict/helpsphinx/gyexamples/plot_where.html). You can run it your machine.

**Comment by waiyujack** [comment]
Thanks xadupre! I think my results are quite consistent with what you've got.

![where_comparison](https://user-images.githubusercontent.com/46226344/101529307-4f145480-3988-11eb-921a-a96e12c38fb8.PNG)

**Comment by waiyujack** [comment]
Hi @xadupre 

Can I ask you one more question? Do you know if there is any onnx function which will do mapping (i.e. categorical mapping) according to a dictionary? An example like below:

```
mapping_dict = {'col_1':{1:1,2:1}, 'col_2':{1:1,2:2}}
df = pd.DataFrame(data= {'col_1':[1,2,1], 'col_2':[1,2,1]})
print(df.replace(mapping_dict))
```
Outputs:
```
   col_1  col_2
0      1      1
1      1      2
2      1      1
```

**Comment by xadupre** [comment]
I think a LabelEncoder is what you are looking for: https://github.com/onnx/onnx/blob/master/docs/Operators-ml.md#ai.onnx.ml.LabelEncoder.

**Comment by waiyujack** [comment]
> I think a LabelEncoder is what you are looking for: https://github.com/onnx/onnx/blob/master/docs/Operators-ml.md#ai.onnx.ml.LabelEncoder.

@xadupre Thanks! I think that's the thingI am looing for. Do you know if there is an example for the function?

I am trying the same function implemented in skl2onnx.algebra.onnx_ops.OnnxLabelEncoder but not so sure how I can adapt your code to the example above.

**Comment by xadupre** [comment]
I guess you want to add it after operator Where?

**Comment by waiyujack** [comment]
@xadupre 
I think I am almost there with the following mapping {0:1,1:2} but got an error at the end saying 
`NotImplemented: [ONNXRuntimeError] : 9 : NOT_IMPLEMENTED : Could not find an implementation for the node La_LabelEncoder:LabelEncoder(2)`
Do you know the reason for that?

Code:
```
# **Imports**

import numpy as np
from sklearn.base import BaseEstimator, TransformerMixin
from sklearn.pipeline import make_pipeline
from onnxruntime import InferenceSession
from skl2onnx import convert_sklearn, to_onnx, wrap_as_onnx_mixin
from skl2onnx.common.data_types import FloatTensorType
from skl2onnx.algebra.onnx_operator_mixin import OnnxOperatorMixin
from skl2onnx.algebra.onnx_ops import OnnxLessOrEqual,OnnxLabelEncoder
from skl2onnx.common.data_types import guess_numpy_type

# **Simple function to check the converted model works fine.**

def predict_with_onnxruntime(onx, X):
    sess = InferenceSession(onx.SerializeToString())
    input_name = sess.get_inputs()[0].name
    res = sess.run(None, {input_name: X.astype(np.float32)})
    return res[0]

# **Custom transformer to perform mapping (NOT YET COMPLETE see last line)**

class CustomOpTransformer(BaseEstimator, TransformerMixin,
                          OnnxOperatorMixin):

    def __init__(self):
        BaseEstimator.__init__(self)
        TransformerMixin.__init__(self)

    def fit(self, X, y=None):
        return self

    def transform(self, X):
        mapping_dict = {0:1,1:2}
        return np.vectorize(mapping_dict.get)(X)

    def onnx_shape_calculator(self):
        def shape_calculator(operator):
            operator.outputs[0].type = operator.inputs[0].type
            print(operator.inputs[0].type)
        return shape_calculator
    
    
    # Mapping using OnnxLabelEncoder
    def to_onnx_operator(self, inputs=None, outputs=('Y', )):
        if inputs is None:
            raise RuntimeError("Parameter inputs should contain at least "
                               "one name.")
        i0 = self.get_inputs(inputs, 0)
        #dtype = guess_numpy_type(i0.type)
        return OnnxLabelEncoder(
                i0
                , keys_floats = [0.,1.]
                , values_floats = [1.,2.]
                , op_version= 12
                , output_names=outputs[0]
                )

# **Try to implement the convertor in a pipeline and test the function**

X = np.array([[0, 0], [1, 1]])#.astype(np.int64)
tr = make_pipeline(CustomOpTransformer())
tr.fit(X)

# Original matrix
print('Original matrix')
print(X)

# sklearn pipeline
print('Transform using np.where')
print(tr.transform(X))

# Run in onnx
print('Transform using to_onnx_operator')
onx = to_onnx(tr, X.astype(np.float32), target_opset=12)
print(predict_with_onnxruntime(onx, X))
```

**Comment by xadupre** [comment]
Which version of onnxruntime are you using?

**Comment by waiyujack** [comment]
> Which version of onnxruntime are you using?

@xadupre 

It's 1.3.0

Ah very interestingly, if you change the mapping output values to int64s it would work but it wouldn't work if the output values are of type floats.

Working example:
```
node = OnnxLabelEncoder('X', keys_floats = [0.,1.], values_int64s = [1,2])
X = np.array([[1,2],[1,3]]).astype(np.float32)
model_def = node.to_onnx({'X': X})

def predict_with_onnxruntime(onx, X):
    sess = InferenceSession(onx.SerializeToString())
    input_name = sess.get_inputs()[0].name
    res = sess.run(None, {input_name: X.astype(np.float32)})
    return res[0]
Y = predict_with_onnxruntime(model_def, X)
print(Y)

```
**Non-example with same error:**
```
node = OnnxLabelEncoder('X', keys_floats = [0.,1.], values_floats = [1.,2.])
X = np.array([[1,2],[1,3]]).astype(np.float32)
model_def = node.to_onnx({'X': X})

def predict_with_onnxruntime(onx, X):
    sess = InferenceSession(onx.SerializeToString())
    input_name = sess.get_inputs()[0].name
    res = sess.run(None, {input_name: X.astype(np.float32)})
    return res[0]

Y = predict_with_onnxruntime(model_def, X)
print(Y)
```

**Comment by waiyujack** [comment]
I have updated to 1.5.2 but i still got the error when I used **values_floats**  instead of **values_int64s** .

Here are my package versions:
```
import onnx
import onnxruntime
import skl2onnx
print(onnx.__version__)
print(onnxruntime.__version__)
print(skl2onnx.__version__)
1.8.0
1.5.2
1.7.1
```

**Comment by xadupre** [comment]
Sorry for the delay. onnxruntime does not implement the whole ONNX specifications to reduce binary size. It is possible to keep int64 as value type for the LabelEncoder and then cast the output into float?

**Comment by waiyujack** [comment]
Thanks for getting back @xadupre! 
I guess I can output it to integers for LabelEncoder and then transform into float using another operators.

**Comment by xadupre** [comment]
I recently created a class which checks if X belongs to one interval among a list of intervals and returns the associated weight (or label). The converter uses a Tree to compute the output and is more efficient with a high number of intervals. See exemple [WOETransformer](https://github.com/onnx/sklearn-onnx/blob/master/docs/tutorial/plot_woe_transformer.py).

Let me know if I can close the issue.

**Comment by xadupre** [comment]
Closing the issue, feel free to reopen it.
